hub/api/pt.py

```diff
@@ -0,0 +1,208 @@
+import os
+import pickle
+import numpy as np
+from itertools import repeat
+from collections import defaultdict
+from hub.util.keys import get_index_map_key
+from hub.core.chunk_engine.chunker import join_chunks
+from hub.core.chunk_engine.read import read_tensor_meta
+from multiprocessing import shared_memory, resource_tracker
+from pathos.pools import ProcessPool
+
+
+# TODO make this use shared memory to make on the fly transforms faster. Currently using transform slows us down by 10x
+def transform_data(transform, data):
+    return transform(data) if transform else data
+
+
+def shared_memory_clear(chunk_set):
+    """Checks if an existing SharedMemory exists with for any chunk in chunk_set and clears it"""
+    for chunk_name in chunk_set:
+        try:
+            shm = shared_memory.SharedMemory(name=chunk_name)
+            shm.close()
+            shm.unlink()
+        except:
+            pass
+
+
+def read_chunks(chunk_path):
+    remove_shared_memory_from_resource_tracker()
+    out = _hub_storage_provider[chunk_path]
+    shm = shared_memory.SharedMemory(
+        create=True, size=len(out), name=chunk_path.split("/")[-1]
+    )
+    shm.buf[:] = out
+    shm.close()
+    return
+
+
+def remove_shared_memory_from_resource_tracker():
+    """Monkey-patch multiprocessing.resource_tracker so SharedMemory won't be tracked
+    More details at: https://bugs.python.org/issue38119
+    """
+
+    def fix_register(name, rtype):
+        if rtype == "shared_memory":
+            return
+        return resource_tracker._resource_tracker.register(self, name, rtype)
+
+    def fix_unregister(name, rtype):
+        if rtype == "shared_memory":
+            return
+        return resource_tracker._resource_tracker.unregister(self, name, rtype)
+
+    resource_tracker.register = fix_register
+    resource_tracker.unregister = fix_unregister
+    if "shared_memory" in resource_tracker._CLEANUP_FUNCS:
+        del resource_tracker._CLEANUP_FUNCS["shared_memory"]
+
+
+# TODO fix warnings at the end of iteration
+def _to_pytorch(dataset, transform=None, workers=1):
+    try:
+        import torch
+    except ModuleNotFoundError:
+        raise Exception  # TODO proper exception
+        # raise ModuleNotInstalledException("torch")
+    global torch
+    global _hub_storage_provider # global to pass to processes, can't serialize it and send as argument
+
+    # TODO, apparently boto3.client isn't safe for multiprocessing https://github.com/boto/boto3/pull/2848/files
+    # could it be working here as we're only reading data?
+    _hub_storage_provider = dataset.provider
+    return TorchDataset(dataset, transform, workers)
+
+
+class TorchDataset:
+    def __init__(self, dataset, transform=None, workers=1):
+        self.dataset = dataset  # TODO disable the memory cache
+        self.transform = transform
+        self.workers = workers
+        self.storage = self.dataset.provider
+        self.pool = ProcessPool(nodes=workers)
```

Comment:
For I/O jobs, usually, a thread pool is more useful, but it wasn't the case in my experiments, not sure why.
Moreover setting the number of nodes to be 2-3x the number of vCPUs has given unexpected speed boosts.